### ISSUE TITLE
Get return type of interface function

### ISSUE THREAD

**Issue by pie6k** [Microsoft/TypeScript#16372]
I've got object that has some function that returns various types. I have another object that is aware of the first object, and I need to know return type of function from first object

Example:

I've got first object like: 

```
const tokens = {
  // key below is name of token
  month: {
    possibleValues: ['jan', 'feb'], // etc...
    filter: (monthName: string): number => {
      return 12; // will return number 1-12
    }, 
  },
  weekday: {
    possibleValues: ['mon', 'tue'], // etc
    filter: (monthName: string): Date => {
      return new Date(); // will return Date of weekday
    },
  },
};

// as seen above - filter of 'month' always returns number and filter of 'weekday' returns Date

const parser: Parser<typeof tokens> {
  parseWeekday: ({ weekday, month }) => {
    // weekday should be detected to be Date type
    // month should be detected to be number type
  }
}
```

Is that possible?

**Comment by mhegazy** [comment]
duplicate of https://github.com/Microsoft/TypeScript/issues/6239, tracked by https://github.com/Microsoft/TypeScript/issues/6606

**Comment by andy-ms** [comment]
You could try it the otherway around, by `tokens` in terms of the other type:
```ts
interface S {
    weekday: Date;
    month: number;
}

type Tokens = {
    [K in keyof S]: {
        possibleValues: string[],
        filter(s: string): S[K],
    }
}

const tokens: Tokens = {
    // key below is name of token
    month: {
        possibleValues: ['jan', 'feb'], // etc...
        filter: (monthName: string): number => {
            return 12; // will return number 1-12
        },
    },
    weekday: {
        possibleValues: ['mon', 'tue'], // etc
        filter: (monthName: string): Date => {
            return new Date(); // will return Date of weekday
        },
    },
};

const parser = {
    parseWeekday: ({ weekday, month }: S) => {
    },
}
```

Not sure how we're supposed to infer the parameter types of `parseWeekday` without an annotation though!

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
